x/btcstaking/keeper/msg_server.go

```diff
@@ -466,6 +470,14 @@ func (ms msgServer) AddCovenantSigs(goCtx context.Context, req *types.MsgAddCove
 		params,
 	)
 
+	// at this point, the covenant signatures are verified and are not duplicated.
+	// Thus, we can safely consider this message as refundable
+	// NOTE: currently we refund tx fee for covenant signatures even if the BTC
```

Comment:
hmm this is not entirely true tbh. We have check for quorum super early in the handler:
```
	if btcDel.HasCovenantQuorums(params.CovenantQuorum) {
		ms.Logger(ctx).Debug("Received covenant signature after achieving quorum", "covenant pk", req.Pk.MarshalHex())
		return &types.MsgAddCovenantSigsResponse{}, nil
	}
```

so this won't be refunded. And refunding it would require a bit of refactor.

Either way I think we cannot refund such case, as then covenant committe member could send his signature forever and bloating chain with it.

Reply:
Hmm I see. So this won't get refunded in the current impl

I actually think we should refund when covenant quorum is reached. Otherwise for each BTC delegation the covenant member has a rough probability of `(covenant size - covenant quorum) / covenant size` to pay some tx fee, so running a covenant emulator won't be free. But the goal of refunding is to ensure running covenant emulator does not incur tx fee.

> Either way I think we cannot refund such case, as then covenant committe member could send his signature forever and bloating chain with it.

Does this attack vector exist? for each BTC delegation a covenant member can only send 1 valid covenant signature, and subsequent ones will be rejected and won't get refunded ([ref](https://github.com/babylonlabs-io/babylon/blob/6e5210d0eacca6570860152d92ec58f4da7c8cf3/x/btcstaking/keeper/msg_server.go#L354-L358)).

Reply:
oh you are right. 

So if we reject duplicates early and not refund them this is fine.

Though handler need refactor either way, as we should refund signatures after quorum only if they are valid and pass all checks.